neuralmonkey/dataset.py

```diff
@@ -0,0 +1,556 @@
+    """
+
+    def __init__(self,
+                 name: str,
+                 iterators: Dict[str, Callable[[], Iterator]],
+                 outputs: Dict[str, Tuple[str, Writer]] = None,
+                 lazy: bool = False,
+                 buffer_size: int = None,
+                 shuffled: bool = False) -> None:
+        """Construct a new instance of the dataset class.
+
+        Do not call this method from the configuration directly. Instead, use
+        the `from_files` function of this module.
+
+        The dataset iterators are provided through factory functions, which
+        return the opened iterators when called with no arguments.
+
+        Arguments:
+            name: The name for the dataset.
+            iterators: A series-iterator generator mapping.
+            lazy: If False, load the data from iterators to a list and store
+                the list in memory.
+            buffer_size: If lazy, use this buffer size for pre-loading data.
+            shuffled: Whether to shuffle the buffer during batching.
+        """
+        self.name = name
+        self.iterators = iterators
+        self.outputs = outputs
+        self.lazy = lazy
+        self.buffer_size = buffer_size
+        self.shuffled = shuffled
+        self.length = None
+
+        if not self.lazy:
+            data = {s_name: list(it())
+                    for s_name, it in self.iterators.items()}
+
+            # Check whether all loaded series have the same length
+            length_dict = {
+                s_name: len(s_data) for s_name, s_data in data.items()}
+            if len(set(length_dict.values())) > 1:
+                raise ValueError("Lengths of data series do not match: {}"
+                                 .format(str(length_dict)))
+
+            self.length = next(iter(length_dict.values()))
+            self.iterators = {
+                s_name: lambda n=s_name: iter(data[n])  # type: ignore
+                for s_name in self.iterators}
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            The length of the dataset. If the dataset is lazy, returns None.
+        """
+        if self.lazy:
+            raise NotImplementedError("Querying the len of a lazy dataset.")
+        assert self.length is not None
+        return self.length
+
+    @property
+    def series(self) -> List[str]:
+        return list(sorted(self.iterators.keys()))
+
+    def has_series(self, name: str) -> bool:
+        """Check if the dataset contains a series of a given name.
+
+        Arguments:
+            name: Series name
+
+        Returns:
+            True if the dataset contains the series, False otherwise.
+        """
+        return name in self.iterators
+
+    def get_series(self, name: str) -> Iterator:
+        """Get the data series with a given name.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            The data series.
```

Comment:
Documentation:
Isn't the function returning an data_series iterator? I would include it in the documentation if so (for clarity).

Reply:
Rephrased:

```
Returns:
    A freshly initialized iterator over the data series. 
```